### ISSUE TITLE
datatable: cell editor crashes 

### ISSUE THREAD

**Issue by KarlKnapp** [primefaces/primefaces#3437]
![datatableundefined](https://user-images.githubusercontent.com/30590617/37151341-5cc5f7b8-22d5-11e8-86c3-fc02442c8e5a.png)

## 1) Environment
- PrimeFaces version: 6.2

## 2) Expected behavior
No exception when using cell editor 
...

## 3) Actual behavior
java.lang.NumberFormatException: For input string: "undefined"
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.lang.Integer.parseInt(Integer.java:580)
	at java.lang.Integer.parseInt(Integer.java:615)
	at org.primefaces.component.datatable.feature.CellEditFeature.encode(CellEditFeature.java:38)
	at org.primefaces.component.datatable.DataTableRenderer.encodeEnd(DataTableRenderer.java:87)
	at javax.faces.component.UIComponentBase.encodeEnd(UIComponentBase.java:920)
	at javax.faces.component.UIComponent.encodeAll(UIComponent.java:1863)
	at com.sun.faces.context.PartialViewContextImpl$PhaseAwareVisitCallback.visit(PartialViewContextImpl.java:582)
	at com.sun.faces.component.visit.PartialVisitContext.invokeVisitCallback(PartialVisitContext.java:183)
	at org.primefaces.component.api.UIData.visitTree(UIData.java:850)

## 4) Steps to reproduce

On Primefaces Showcase for Datatable/Celleditor:

Select the last cell of the last row.
tab.
click anywhere on the page twice.

**Comment by alissongodoi** [comment]
I am having this issue in production.
Is there any workaround?
Regards.

**Comment by grantsmith1969** [comment]
We're having the same problem. We have had to remove all p:cellEditor components and just use a valueChangeListener inside p:inputText and a p:ajax child within that. Something like this:

```
<p:column>
  <p:inputText value="#{completion.donePercentage}"
                       valueChangeListener="#{completionScreen.onCellEdit}" style="width: 80%">
    <p:ajax global="false"/>
  </p:inputText>
</p:column>
```


Our listener class looks something like this:

```
public void onCellEdit(ValueChangeEvent event) {
        if (event.getPhaseId() != PhaseId.INVOKE_APPLICATION) {
            event.setPhaseId(PhaseId.INVOKE_APPLICATION);
            event.queue();
            return;
        }
        // if you need the old and new entered values:
        Object oldValue = event.getOldValue();
        Object newValue = event.getNewValue();
        // If you need the whole row:
        Completion selectedCompletion = Faces.evaluateExpressionGet("#{completion}");
        // persist logic:
        projectDAO.saveCompletion(selectedCompletion);
 }
```

**Comment by marcelocaser** [comment]
thank you! i have the same problem!!

**Comment by huseyinturkmenoglu** [comment]
I tried this issue. But I don't see any issue from showcase. 

(https://www.primefaces.org/showcase/ui/data/datatable/edit.xhtml)

Could you send sample code (https://github.com/primefaces/primefaces-test) and gif/video ? 

Greetings,

**Comment by tandraschko** [comment]
@HuseyinTurkmenoglu It would be great if you could set the "can't replicate" label for such issues.

**Comment by FlavioSantoro92** [comment]
@HuseyinTurkmenoglu in the showcase now works normally because the version of PF is 6.3-SNAPSHOT.
This bug is present in the version 6.2 and lower. In fact, few days ago, when the version on the showcase was 6.2, I tried to go on the last CellEdit of a DataTable, I pressed Tab and double-click anywhere and the whole app crashed.

**Comment by KarlKnapp** [comment]
I'm not sure if this is related, but if I remember correctly in the previous version of the showcase the last cell was an input field, in the current version it's a selectoOneMenu.

**Comment by ReedElliott** [comment]
I was able to reproduce this issue in my application running on PrimeFaces 6.2.3.   I believe the issue is associated with the PrimeFaces DataTable component posting a "_cellInfo=undefined,-1" parameter when the user selects the editable field.   The weird part for my case is that it doesn't do this on the fields in the first row of the table, but it does do it on the fields in the second row.   Note: I use ajax to add the 1st and the 2nd row to the table.

**Comment by FlavioSantoro92** [comment]
Even on the version 6.2.5 in the showcase the bug is still present.
To reproduce:

Go here: https://www.primefaces.org/showcase/ui/data/datatable/edit.xhtml
In the second datatable, select the last cell, press tab and double click anywhere.
Still crash.

**Comment by ReedElliott** [comment]
If you open the Network tab on the Firefox Developer tools, you will see the following param being sent to the server in the post that occurs when you perform the double click as described by @FlavioSantoro92 above.

- form:cars2_cellInfo: undefined,-1

I believe that this param is causing the following error on the server:

> java.lang.NumberFormatException: For input string: "undefined"
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.lang.Integer.parseInt(Integer.java:580)
	at java.lang.Integer.parseInt(Integer.java:615)
	at org.primefaces.component.datatable.feature.CellEditFeature.encode(CellEditFeature.java:38)
	at org.primefaces.component.datatable.DataTableRenderer.encodeEnd(DataTableRenderer.java:87)

**Comment by melloware** [comment]
I have reported this through my PRO account so hopefully it will get resolved for PF 6.2.6.

**Comment by marcdas** [comment]
Este defeito ja foi corrigido?
Estou com o mesmo problema e não sei resolver.

**Comment by marcdas** [comment]
Has this been fixed yet?
I have the same problem and I do not know how to solve it.

**Comment by melloware** [comment]
It was fixed in PF 6.2.6 so if you are an Elite member you can download it.  Or you can build the JAR from the GitHub source!

**Comment by marcdas** [comment]
melloware, thank you very much for your attention.
I do not have so much experience in this procedure.
When doing the clone in the eclipse appeared the problem below:
If not ask too much, can anyone give me some guidance this.

- In the marketplace entries found to handle maven-jsf-plugin: 1.3.3-SNAPSHOT: generate-components in Eclipse.
- In the marketplace entries found to handle maven-jsf-plugin: 1.3.3-SNAPSHOT: generate-facelets-taglib in Eclipse.
- In the marketplace entries found to handle maven-jsf-plugin: 1.3.3-
SNAPSHOT: generate-faces-config in Eclipse.
- In the marketplace entries found to handle xslt-maven-plugin: 1.0: transform in Eclipse.

**Comment by melloware** [comment]
Here are the instructions you need to build from source: https://github.com/primefaces/primefaces/wiki/Building-From-Source

**Comment by Alireza-Jamali** [comment]
I had the same problem @marcdas had, you've to install `maven-jsf-plugin` into your local repository like this: 
```
mvn install:install-file -Dfile=<path-to-file> -DgroupId=<group-id> \
    -DartifactId=<artifact-id> -Dversion=<version> -Dpackaging=<packaging>

```
 and edit its version in primefaces pom file, 
guy [here ](https://forum.primefaces.org/viewtopic.php?f=3&t=39299&sid=c4011a3d6e5391a81d8306208fd6648e)had the same problem.

**Comment by stochmalm** [comment]
For anybody who can't wait for the 6.3 release here's a hack with this fix. Just add it to <script> in your template.

     $(document).ready(function () {
            //Fix for primefaces bug: https://github.com/primefaces/primefaces/issues/3437
            PrimeFaces.widget.DataTable.prototype.showCurrentCell = function (cell) {
                var $this = this;

                if (this.currentCell) {
                    if (this.cfg.saveOnCellBlur)
                        this.saveCell(this.currentCell);
                    else if (!this.currentCell.is(cell))
                        this.doCellEditCancelRequest(this.currentCell);
                }
                if (cell && cell.length) {
                    this.currentCell = cell;

                    var cellEditor = cell.children('div.ui-cell-editor'),
                        displayContainer = cellEditor.children('div.ui-cell-editor-output'),
                        inputContainer = cellEditor.children('div.ui-cell-editor-input'),
                        inputs = inputContainer.find(':input:enabled'),
                        multi = inputs.length > 1;

                    cell.addClass('ui-state-highlight ui-cell-editing');
                    displayContainer.hide();
                    inputContainer.show();
                    inputs.eq(0).focus().select();

                    //metadata
                    if (multi) {
                        var oldValues = [];
                        for (var i = 0; i < inputs.length; i++) {
                            var input = inputs.eq(i);

                            if (input.is(':checkbox')) {
                                oldValues.push(input.val() + "_" + input.is(':checked'));
                            } else {
                                oldValues.push(input.val());
                            }
                        }

                        cell.data('multi-edit', true);
                        cell.data('old-value', oldValues);
                    } else {
                        cell.data('multi-edit', false);
                        cell.data('old-value', inputs.eq(0).val());
                    }

                    //bind events on demand
                    if (!cell.data('edit-events-bound')) {
                        cell.data('edit-events-bound', true);

                        inputs.on('keydown.datatable-cell', function (e) {
                            var keyCode = $.ui.keyCode,
                                shiftKey = e.shiftKey,
                                key = e.which,
                                input = $(this);

                            if (key === keyCode.ENTER || key == keyCode.NUMPAD_ENTER) {
                                $this.saveCell(cell);

                                e.preventDefault();
                            } else if (key === keyCode.TAB) {
                                if (multi) {
                                    var focusIndex = shiftKey ? input.index() - 1 : input.index() + 1;

                                    if (focusIndex < 0 || (focusIndex === inputs.length) || input.parent().hasClass('ui-inputnumber') || input.parent().hasClass('ui-helper-hidden-accessible')) {
                                        $this.tabCell(cell, !shiftKey);
                                    } else {
                                        inputs.eq(focusIndex).focus();
                                    }
                                } else {
                                    $this.tabCell(cell, !shiftKey);
                                }

                                e.preventDefault();
                            } else if (key === keyCode.ESCAPE) {
                                $this.doCellEditCancelRequest(cell);
                                e.preventDefault();
                            }
                        })
                            .on('focus.datatable-cell click.datatable-cell', function (e) {
                                $this.currentCell = cell;
                            });
                    }
                } else {
                    this.currentCell = null;
                }
            }
            //showCurrentCell END
        });

**Comment by mohammediqraminamdar** [comment]
i used above script but not work 
how to solve this issue or  which primeFaces version that s resolved ?

**Comment by melloware** [comment]
Its fixed in 6.2.6+ or 7.0+
